### PR TITLE
wrapper/tencentcloudkms: add tencent cloud kms wrapper

### PR DESCRIPTION
**Pull request by likexian**
Hello,

This PR add supports of TencentCloud KMS.

## About TencentCloud
TencentCloud is a secure, reliable and high-performance cloud compute service provided by Tencent.  It is the 2st largest Cloud Provider in China.

### CODE REVIEW

**Comment by likexian** [comment]
Hello @jefferai 
Would you please have a review?

File: wrappers/tencentcloudkms/tencentcloudkms.go [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+	wrapping "github.com/hashicorp/go-kms-wrapping"
+	"github.com/tencentcloud/tencentcloud-sdk-go/tencentcloud/common"
+	"github.com/tencentcloud/tencentcloud-sdk-go/tencentcloud/common/profile"
+	kms "github.com/tencentcloud/tencentcloud-sdk-go/tencentcloud/kms/v20190118"
+)
+
+// These constants contain the accepted env vars; the Vault one is for backwards compat
+const (
+	EnvTencentCloudKMSWrapperKeyID   = "TENCENTCLOUDKMS_WRAPPER_KEY_ID"
+	EnvVaultTencentCloudKMSSealKeyID = "VAULT_TENCENTCLOUDKMS_SEAL_KEY_ID"
```
Comment (jefferai) [inline wrappers/tencentcloudkms/tencentcloudkms.go]: This is unnecessary; it never existed in Vault.

File: wrappers/tencentcloudkms/tencentcloudkms.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+// * Instance metadata role (access key and secret key)
+func (k *Wrapper) SetConfig(config map[string]string) (map[string]string, error) {
+	if config == nil {
+		config = map[string]string{}
+	}
+
+	switch {
+	case os.Getenv(EnvTencentCloudKMSWrapperKeyID) != "":
+		k.keyID = os.Getenv(EnvTencentCloudKMSWrapperKeyID)
+	case os.Getenv(EnvVaultTencentCloudKMSSealKeyID) != "":
```
Comment (jefferai) [inline wrappers/tencentcloudkms/tencentcloudkms.go]: Unnecessary, please remove

File: wrappers/tencentcloudkms/tencentcloudkms.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+// Wrapper is a wrapper that uses TencentCloud KMS
+type Wrapper struct {
+	accessKey    string
+	secretKey    string
+	sessionToken string
+	region       string
+
+	keyID        string
+	currentKeyID *atomic.Value
```
Comment (jefferai) [inline wrappers/tencentcloudkms/tencentcloudkms.go]: I'm not sure about this currentKeyID mechanism. If this value can be changed on every encrypt operation, then it seems like any time you do a KeyID() call, this value might have been overwritten multiple times. In other words, I'm not sure it's meaningful to display the current actual value as opposed to the original alias.

Comment (likexian) [reply wrappers/tencentcloudkms/tencentcloudkms.go]: Hello @jefferai , thanks for your reviews.

For tencentcloud kms, every encrypt operation will returns a unique `encrypt key id`, which I stores it to `currentKeyID`, this `encrypt key id` is totally different from the `kms key id` that setting in the configure. When do a `KeyID()` call, it just returns the `encrypt key id`, it is a read function.

**Comment by likexian** [comment]
I have make the requested changes, the tests passed.
Please help have a new review.

```
=== RUN   TestAccTencentCloudKMSWrapper_Lifecycle
--- PASS: TestAccTencentCloudKMSWrapper_Lifecycle (0.29s)
=== RUN   TestTencentCloudKMSWrapper
--- PASS: TestTencentCloudKMSWrapper (0.00s)
=== RUN   TestTencentCloudKMSWrapper_Lifecycle
--- PASS: TestTencentCloudKMSWrapper_Lifecycle (0.00s)
PASS
ok      github.com/hashicorp/go-kms-wrapping/wrappers/tencentcloudkms   0.297s
```

File: wrappers/tencentcloudkms/tencentcloudkms.go (hunk 3 above)
Comment (jefferai) [reply wrappers/tencentcloudkms/tencentcloudkms.go]: That's my point though. The KeyID call will return different values over time, but that is not necessarily the value you're expecting. Let's say you have multiple goroutines using this wrapper. Routine A does an encrypt, then B does an encrypt. Then A reads KeyID but what it gets back isn't the key ID that corresponded to the encrypt call it performed, but instead it gets the key ID corresponding to B's encryption call.

Comment (jefferai) [reply wrappers/tencentcloudkms/tencentcloudkms.go]: Are you sure _every_ encrypt operation returns a _unique_ key ID? As opposed to the key changing over time due to e.g. periodic rotation?

Comment (likexian) [reply wrappers/tencentcloudkms/tencentcloudkms.go]: Yes, every encrypt operation uses a golbal `kms key id` and a unique `encrypt key id`. You may image that every encrypt operation uses a key (`kms key id`) and a salt (`encrypt key id`).

Comment (likexian) [reply wrappers/tencentcloudkms/tencentcloudkms.go]: > That's my point though. The KeyID call will return different values over time, but that is not necessarily the value you're expecting. Let's say you have multiple goroutines using this wrapper. Routine A does an encrypt, then B does an encrypt. Then A reads KeyID but what it gets back isn't the key ID that corresponded to the encrypt call it performed, but instead it gets the key ID corresponding to B's encryption call.

So, what is your suggestions? remove the function `KeyID()` ?

Comment (likexian) [reply wrappers/tencentcloudkms/tencentcloudkms.go]: > Are you sure _every_ encrypt operation returns a _unique_ key ID? As opposed to the key changing over time due to e.g. periodic rotation?

Sorry, I correct what I said: tencentcloud always returns the fixed `kms key id`, what I said below is wrong.

File: wrappers/tencentcloudkms/tencentcloudkms.go [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+		if keyInfo.Response.KeyMetadata == nil || keyInfo.Response.KeyMetadata.KeyId == nil {
+			return nil, fmt.Errorf("no key information return")
+		}
+
+		k.currentKeyID.Store(*keyInfo.Response.KeyMetadata.KeyId)
+		k.client = client
+	}
+
+	wrappingInfo := make(map[string]string)
+	wrappingInfo["retgion"] = k.region
```
Comment (jefferai) [inline wrappers/tencentcloudkms/tencentcloudkms.go]: typo

File: wrapper.go [hunk 5]
```diff
@@ -8,17 +8,18 @@ import (
...
-	Test           = "test-auto"
+	AEAD            = "aead"
+	Shamir          = "shamir"
+	PKCS11          = "pkcs11"
+	AliCloudKMS     = "alicloudkms"
+	AWSKMS          = "awskms"
+	GCPCKMS         = "gcpckms"
+	AzureKeyVault   = "azurekeyvault"
+	OCIKMS          = "ocikms"
+  HuaweiCloudKMS = "huaweicloudkms"
```
Comment (jefferai) [inline wrapper.go]: Formatting changed for some reason

File: wrappers/tencentcloudkms/tencentcloudkms.go (hunk 4 above)
Comment (likexian) [reply wrappers/tencentcloudkms/tencentcloudkms.go]: Sorry!

File: wrapper.go (hunk 5 above)
Comment (likexian) [reply wrapper.go]: Formatting changed due to github conflict merge, fixed.
The tests of HuaweiCloudKMS never passed, I also fixed it.

**Comment by likexian** [comment]
> I'm going to let it go this time but in the future, please do not lint/adjust things that are not in scope of the PR.
> 
> LGTM!

Sorry, got it.
I fixed it because the global `go test` failed.

**Pull request merged**